pkg/reconciler/extension/extension.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2022 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package extension
+
+import (
+	"context"
+
+	autoscalingv1alpha1 "knative.dev/serving/pkg/apis/autoscaling/v1alpha1"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// ServingExtension enables pluggable extensive features added into the controller
+type ServingExtension interface {
+	// TransformRevision transforms the existing revision.
+	TransformRevision(*v1.Revision) *v1.Revision
+	// PostRevisionReconcile implements a custom reconcile loop the after the revision is created.
+	PostRevisionReconcile(context.Context, *autoscalingv1alpha1.PodAutoscaler, *autoscalingv1alpha1.PodAutoscaler) error
+	// PostConfigurationReconcile implements a custom reconcile loop after the configuration is created.
+	PostConfigurationReconcile(context.Context, *v1.Service, *v1.Configuration) error
+	// TransformService transforms the existing service.
+	TransformService(*v1.Service) *v1.Service
```

Comment:
Which service is being transformed - and what transformations are you doing?

Reply:
The service contains the traffic information. the only change I am about to do is to fill in the traffic split based on each stage, till it reaches the end of the transition.
Currently, the traffic information is directly from the user input. Instead, I use this transformed service with the traffic information for each stage(this transformed service is not applied to the kubernetes cluster), as the input for the `route`.